executor/index_lookup_join.go

```diff
@@ -129,6 +129,25 @@ type innerWorker struct {
 
 // Open implements the Executor interface.
 func (e *IndexLookUpJoin) Open(ctx context.Context) error {
+	// Be careful, very dirty hack in this line!!!
+	// IndexLookUpJoin need to rebuild executor (the dataReaderBuilder) during
+	// executing. However `executor.Next()` is lazy evaluation when the RecordSet
+	// result is drained.
+	// Lazy evaluation means the saved session context may change during executor's
+	// building and its running.
+	// A specific sequence for example:
+	//
+	// e := buildExecutor()   // txn at build time
+	// recordSet := runStmt(e)
+	// session.CommitTxn()    // txn closed
+	// recordSet.Next()
+	// e.dataReaderBuilder.Build() // txn is used again, which is already closed
+	//
+	// The trick here is `getStartTS` will cache start ts in the dataReaderBuilder,
+	// so even txn is destroyed later, the dataReaderBuilder could still use the
+	// cached start ts to construct DAG.
+	e.innerCtx.readerBuilder.getStartTS()
```

Comment:
I think a better way is to reuse the inner table reader, not rebuilding it.

Reply:
At my first thought, reuse table reader involves state management.
Maintain the state is not a trivial work, there are many channels, goroutines... 
and it tend to cause resource leak.
Rebuild and discard table reader makes the code simpler.

Considering this special case now, I think it does have some advantages if we possibility gather more things during building stage and avoid re-build again in the executing stage.

Reply:
Anyway, let's focus on the issue resolved in this commit